### ISSUE TITLE
Fragmentation issue

### ISSUE THREAD

**Issue by susheelbhanu** [pereiramemo/AGS-and-ACN-tools#2]
Hi, 

Thanks for the tool. I've tried it today, but seem to keep running into a "segmentation fault". Please see the error message below:
```
Predicting ORFs ...
FragGeneScanPlusPlus failed
/bioinfo/software/ags.sh: line 431:    62 Segmentation fault      "${fraggenescanplusplus}" -s "${INPUT_FNA}" -o "${ORFS_OUT}" -w 0 -r "${FGSP_TRAIN_DIR}" -t "${TRAIN_FILE_NAME}" -c 1```

I've tried this with a .fna, .faa, and a fasta file, but get the same error. Is this a memory issue?

Thank you,
Susheel

**Comment by susheelbhanu** [comment]
I was able to fix this by giving both the .fna and .faa files, so please ignore. For others who may run into the same issue, here's how I ran it:
```./run_ags.sh B4.fna B4.faa B4_ags_output --verbose t --min_length 120 --nslots 4 --overwrite t --save_complementary_data t```

Thank you!

**Comment by pereiramemo** [comment]
Dear Susheel,

Thanks for your interest. 
I should mention that this segmentation fault issue with FragGeneScanPlusPlus appears to be related to the training file used with certain sequence files. The default training file we have in ags.sh is "illumina_5" (Illumina sequencing reads with about 0.5% error rate). However, this might be too high if you have good quality sequences. If you change the training file (with the --train_file_name parameter) to "illumina_1", you might not get the segmentation fault error, and you should get the same (or almost de same) AGS.

Best, 

Emiliano

**Comment by susheelbhanu** [comment]
Thank you, that's good to know!

**Comment by marschmi** [comment]
Hi there! I've been running AGS with strict quality controlled fastq.gz sequences [as per #1 and the suggestion above] and for ~10% of my samples (in the same experimental group), I'm getting a `136 segmentation fault` (note that the original issue was with 62 segmentation fault). The other 90% seem ok and the samples that fail have the same amount of reads as the ones that worked.

I've been running the following command:
```
# Calculate AGS
/bin/AGS-and-ACN-tools/run_ags.sh \
                        ../../trimmed_QCseqs_123_interleaved.fastq.gz ags_123_output \
                        --min_length 120 \
                        --max_length 180 \
                        --sample_name sample_123 \
                        --verbose t \
                        --overwrite t \
                        --nslots 15 \
                        --train_file_name illumina_1 \
                        --save_complementary_data t
```

which gives the following error:
```
/bioinfo/software/ags.sh: line 431:   136 Segmentation fault      (core dumped) "${fraggenescanplusplus}" -s "${INPUT_FNA}" -o "${ORFS_OUT}" -w 0 -r "${FGSP_TRAIN_DIR}" -t "${TRAIN_FILE_NAME}" -c 1 -p "${NSLOTS}" 2>&1
       137 Done                    | handleoutput
```

I would be grateful for any help with debugging this error!

**Comment by pereiramemo** [comment]
Hi, 

This is an issue observed in FGSpp. Did you try with a different error model (i.e., illumina_5 or illumina_10)?
What is the read length of your sequences?

**Comment by marschmi** [comment]
Hi there, 

Yes, I tried the other error models `illumina_1`, `illumina_5`, and `illumina_10` and they all gave the same segmentation error. This was weird as the samples that failed have only sequences that are 2-5 bases smaller on average (avg of 115-119bp instead of avg of 117-121bp, skewed by small sequences still in the data (the median is >120bp)). I noticed that  `FTrimmed` was always been 0 for my samples, even with `illumina_1` (as I mentioned I used a very strict quality threshold). So for these last 10% of samples, I decided to use `--train_file_name complete`. The data was still within range of the other lanes for the same samples so it seems that all is good now.

I've successfully calculated and now can go to analysis. Thanks for the help!

**Comment by pereiramemo** [comment]
Good to know :)
Cheers, 

Emiliano

**Comment by pereiramemo** [comment]
I just realized that given the read length of your sequences, you should consider lowering the minimum length parameter "--min_length". I would say 110 could be an option. However, the most appropriate number will depend on your read length distribution.

**Comment by marschmi** [comment]
Hi Emiliano, 

I was considering this but felt that I should listen to the emphasis in your paper regarding sequences between 120-180bp. Therefore, I have currently cut them to be 120-180 bp. Though, this has removed 22-30% of the bases in each sample, which originate from samples that have 2.5x10<sup>9</sup> to 4.2x10<sup>9</sup> bases. I can run a few of the samples with `--min_length 110` and see how it modifies the results. As you're the expert in this tool, I welcome your thoughts on this too!

Thank you so much for your input. Your perspective is greatly appreciated!

**Comment by pereiramemo** [comment]
I thought that for the samples that have an average read length lower than 120bp, the number of removed bases could have been greater. If that is not the case, then yes, using a read length between 120 to 180bp should be the best option.  
Just to know: what was the proportion of reads removed?

**Comment by johannalheureux23** [comment]
Hi @pereiramemo I am having a very similar issue when running run.ags and am unsure how to solve it. I've tried using the different error models (illumina_1, illumina_5, and illumina_10) but am getting the same error. I've also tried using different minlengths. 

What's strange is that for one of my samples, the software works however for several others that I've tried, it gives the segmentation error at the Fraggenescan step. Any ideas on how else I can try and solve this? 

Thank you!
![error](https://github.com/pereiramemo/AGS-and-ACN-tools/assets/138055483/aaeb2e59-4dd0-49df-9f64-8cc686a4159c)
